src/Features/Core/Portable/EditAndContinue/Remote/RemoteEditAndContinueServiceProxy.cs

```diff
@@ -0,0 +1,462 @@
+        {
+            _workspace = workspace;
+        }
+
+        public Workspace Workspace => _workspace;
+
+        private sealed class DocumentActiveStatementProviderCallback : IRemoteEditAndContinueService.IDocumentActiveStatementProviderCallback
+        {
+            private readonly DocumentActiveStatementSpanProvider _provider;
+
+            public DocumentActiveStatementProviderCallback(DocumentActiveStatementSpanProvider provider)
+                => _provider = provider;
+
+            /// <summary>
+            /// Remote API.
+            /// </summary>
+            public async Task<ImmutableArray<TextSpan>> GetSpansAsync(CancellationToken cancellationToken)
+            {
+                try
+                {
+                    return await _provider(cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e))
+                {
+                    return ImmutableArray<TextSpan>.Empty;
+                }
+            }
+        }
+
+        private sealed class SolutionActiveStatementProviderCallback : IRemoteEditAndContinueService.ISolutionActiveStatementProviderCallback
+        {
+            private readonly SolutionActiveStatementSpanProvider _provider;
+
+            public SolutionActiveStatementProviderCallback(SolutionActiveStatementSpanProvider provider)
+                => _provider = provider;
+
+            /// <summary>
+            /// Remote API.
+            /// </summary>
+            public async Task<ImmutableArray<TextSpan>> GetSpansAsync(DocumentId documentId, CancellationToken cancellationToken)
+            {
+                try
+                {
+                    return await _provider(documentId, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e))
+                {
+                    return ImmutableArray<TextSpan>.Empty;
+                }
+            }
+        }
+
+        private IEditAndContinueWorkspaceService GetLocalService()
+            => _workspace.Services.GetRequiredService<IEditAndContinueWorkspaceService>();
+
+        public async Task StartDebuggingSessionAsync(Solution solution, CancellationToken cancellationToken)
+        {
+            var client = await RemoteHostClient.TryGetClientAsync(_workspace, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                GetLocalService().StartDebuggingSession(solution);
+                return;
+            }
+
+            await client.RunRemoteAsync(
+                WellKnownServiceHubService.RemoteEditAndContinueService,
+                nameof(IRemoteEditAndContinueService.StartDebuggingSessionAsync),
+                solution,
+                Array.Empty<object>(),
+                callbackTarget: null,
+                cancellationToken).ConfigureAwait(false);
+        }
```

Comment:
i'm a bit surprised we don't create and hold onto a single client here.  feels like that makes sense as that would be what we would talk to for the lifetime of the debugging/enc session

Reply:
RemoteEditAndContinueServiceProxy is just a facade that contains all calls to RemoteEditAndContinueService. Other services work the same way, except the methods are spread out. Imagine we inlined all the methods into their callers. 

This way makes it easier to test serialization.